src/classes/Query.cls

```diff
@@ -1623,6 +1640,11 @@ public class Query {
         public Boolean isUnique = false;
         public Boolean isUpdateable = false;
         public Boolean isWriteRequiresMasterRead = false;
+        public Boolean isSystemContext = false;
```

Comment:
As I elaborated in the comment, you can try implement it without introducing `isSystemContext` in `FieldSetting`.